addons/product_expiry/models/stock_move_line.py

```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import datetime
+
+from odoo import api, fields, models, _
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+    expiration_date = fields.Boolean(related='product_id.expiration_date')
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    def _default_life_date(self):
+        product = self.env['product.product'].browse(self.env.context.get('default_product_id'))
+        if not product.expiration_date:
+            return
+        return fields.Datetime.today() + datetime.timedelta(days=product.life_time)
+
+    expiration_date = fields.Boolean(related='product_id.expiration_date')
+    life_date = fields.Datetime(string='Expiration Date',
+        help='This is the date on which the goods with this Serial Number may'
+        ' become dangerous and must not be consumed.',
+        default=lambda self: self._default_life_date())
+
+    def _action_done(self):
+        super(StockMoveLine, self)._action_done()
+        lines = self.filtered(lambda line: line.exists() and line.life_date)
+        for line in lines:
+            line.lot_id._update_date_values(line.life_date)
```

Comment:
_action_done of stock move line may unlink records in self, here you iterate and work on all stock move line which could lead to tracebacks

i would update the original _action_done to return the records not unlinked
```
return self - ml_to_delete

Reply:
you could also extract the lot creation in the original method
``` python
                            # If a picking type is linked, we may have to create a production lot on
                            # the fly before assigning it to the move line if the user checked both
                            # `use_create_lots` and `use_existing_lots`.
                            if ml.lot_name and not ml.lot_id:
                                lot = self.env['stock.production.lot'].create(
                                    {'name': ml.lot_name, 'product_id': ml.product_id.id, 'company_id': ml.move_id.company_id.id}
                                )
                                ml.write({'lot_id': lot.id})
```
in an helper and override the helper here to complete them?